src/DropdownReferenceSet/components/DropdownReferenceSetContainer.ts

```diff
@@ -0,0 +1,256 @@
+                        label: ""
+                    }
+                });
+            }
+        });
+    }
+
+    private executeOnChangeEvent = () => {
+        const { mxform, mxObject, onChangeEvent, onChangeMicroflow, onChangeNanoflow } = this.props;
+        const context = new mendix.lib.MxContext();
+
+        context.setContext(mxObject.getEntity(), mxObject.getGuid());
+        if (onChangeEvent === "callMicroflow" && onChangeMicroflow) {
+            window.mx.ui.action(onChangeMicroflow, {
+                error: error => window.mx.ui.error(`Error while executing microflow ${onChangeMicroflow}: ${error.message}`), // tslint:disable-line max-line-length
+                origin: mxform,
+                params: {
+                    applyto: "selection",
+                    guids: [ mxObject.getGuid() ]
+                }
+            });
+        } else if (onChangeEvent === "callNanoflow" && onChangeNanoflow.nanoflow) {
+            window.mx.data.callNanoflow({
+                context,
+                error: error => window.mx.ui.error(`Error while executing the onchange nanoflow: ${error.message}`),
+                nanoflow: onChangeNanoflow,
+                origin: mxform
+            });
+        }
+    }
+
+    private retrieveOptions(props: ContainerProps) {
+        const entity = this.props.entityPath.split("/")[1];
+        const { sortAttributes, attribute, entityConstraint, source, sortOrder, microflow, mxObject, nanoflow } = props;
+        const attributeReference = `${props.entityPath}${attribute}`;
+        const options: FetchDataOptions = {
+            attributes: [ attributeReference ],
+            constraint: entityConstraint,
+            entity,
+            guid: mxObject.getGuid(),
+            microflow,
+            mxform: this.props.mxform,
+            nanoflow,
+            sortAttributes: !sortAttributes.length ?
+                [ { name: attribute , sort: "asc" } ] :
+                sortAttributes,
+            sortOrder,
+            source
+        };
+
+        fetchData(options)
+            .then(optionObjects => this.setOptions(optionObjects))
+            .catch(errorMessage => window.mx.ui.error(errorMessage.message));
+    }
+
+    private setOptions = (mendixObjects: mendix.lib.MxObject[]) => {
+        const options: ReferenceOption[] = mendixObjects.map(mxObject => {
+            return {
+                label: mx.parser.formatAttribute(mxObject, this.props.attribute),
+                value: mxObject.getGuid()
+            };
+        });
+
+        this.setState({ options });
+    }
+
+    private setAsyncOptions = (input: string): Promise<{ options: ReferenceOption[] }> => {
+        if (this.props.mxObject) {
+            this.props.mxObject.set(this.props.searchAttribute, input);
+            if (input.length >= this.props.minimumCharacter) {
+                this.retrieveOptions(this.props);
+
+                return Promise.resolve({ options: this.state.options });
```

Comment:
Sure? it should fully async?
retrieveOptions is also async. and should only resolve when both are finished